File: flutter_app/lib/api/api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/app_config.dart';
import '../core/token_storage.dart';

class ApiException implements Exception {
  final String message;
  final int? statusCode;

  ApiException(this.message, {this.statusCode});

  @override
  String toString() => 'ApiException($statusCode): $message';
}

class ApiClient {
  ApiClient(this._tokenStorage);

  final TokenStorage _tokenStorage;

  Uri _buildUri(String route, [Map<String, String>? query]) {
    final normalizedRoute = route.startsWith('/') ? route.substring(1) : route;
    return Uri.parse(AppConfig.baseUrl).replace(
      queryParameters: <String, String>{
        'route': normalizedRoute,
        ...?query,
      },
    );
  }

  Future<Map<String, String>> _headers({bool jsonBody = true}) async {
    final token = await _tokenStorage.readToken();
    final headers = <String, String>{};
    if (jsonBody) {
      headers['Content-Type'] = 'application/json';
    }
    if (token != null && token.isNotEmpty) {
      headers['Authorization'] = 'Bearer $token';
    }
    return headers;
  }

  Future<dynamic> get(String route, {Map<String, String>? query}) async {
    final response = await http.get(
      _buildUri(route, query),
      headers: await _headers(),
    );
    return _decodeResponse(response);
  }

  Future<dynamic> post(String route, {Map<String, dynamic>? body}) async {
    final response = await http.post(
      _buildUri(route),
      headers: await _headers(),
      body: jsonEncode(body ?? <String, dynamic>{}),
    );
    return _decodeResponse(response);
  }

  Future<dynamic> postMultipart(
    String route, {
    required Map<String, String> fields,
    required String fileField,
    required String filePath,
  }) async {
    final request = http.MultipartRequest('POST', _buildUri(route));
    final token = await _tokenStorage.readToken();
    if (token != null && token.isNotEmpty) {
      request.headers['Authorization'] = 'Bearer $token';
    }
    request.fields.addAll(fields);
    request.files.add(await http.MultipartFile.fromPath(fileField, filePath));
    final streamed = await request.send();
    final response = await http.Response.fromStream(streamed);
    return _decodeResponse(response);
  }

  dynamic _decodeResponse(http.Response response) {
    final Map<String, dynamic> payload = jsonDecode(response.body) as Map<String, dynamic>;
    final success = payload['success'] == true;
    if (response.statusCode >= 200 && response.statusCode < 300 && success) {
      return payload['data'];
    }
    throw ApiException(
      (payload['error'] ?? 'Request failed').toString(),
      statusCode: response.statusCode,
    );
  }
}

File: frontend/src/api/apiClient.ts
import { AppConfig } from '../core/appConfig';
import { TokenStorage } from '../core/tokenStorage';
import { asRecord } from '../models/parse';

interface ApiPayload {
  success?: boolean;
  data?: unknown;
  error?: unknown;
}

export interface MultipartFile {
  uri: string;
  name?: string;
  type?: string;
}

export class ApiException extends Error {
  statusCode?: number;

  constructor(message: string, statusCode?: number) {
    super(message);
    this.name = 'ApiException';
    this.statusCode = statusCode;
  }
}

function extractErrorMessage(error: unknown): string {
  if (error instanceof Error && error.message.length > 0) {
    return error.message;
  }
  return String(error ?? 'Request failed');
}

function guessMimeType(filePath: string): string {
  const normalized = filePath.toLowerCase();
  if (normalized.endsWith('.png')) {
    return 'image/png';
  }
  if (normalized.endsWith('.jpg') || normalized.endsWith('.jpeg')) {
    return 'image/jpeg';
  }
  if (normalized.endsWith('.pdf')) {
    return 'application/pdf';
  }
  return 'application/octet-stream';
}

export class ApiClient {
  constructor(private readonly tokenStorage: TokenStorage) {}

  private buildUrl(route: string, query?: Record<string, string>): string {
    const normalizedRoute = route.startsWith('/') ? route.slice(1) : route;
    const params = new URLSearchParams({ route: normalizedRoute });
    if (query !== undefined) {
      Object.entries(query).forEach(([key, value]) => {
        if (value !== '') {
          params.set(key, value);
        }
      });
    }
    return `${AppConfig.baseUrl}?${params.toString()}`;
  }

  private async headers(jsonBody = true): Promise<Record<string, string>> {
    const token = await this.tokenStorage.readToken();
    const headers: Record<string, string> = {};
    if (jsonBody) {
      headers['Content-Type'] = 'application/json';
    }
    if (token !== null && token.length > 0) {
      headers.Authorization = `Bearer ${token}`;
    }
    return headers;
  }

  private parsePayload(rawBody: string): ApiPayload {
    try {
      return asRecord(JSON.parse(rawBody)) as ApiPayload;
    } catch {
      return {
        success: false,
        error: 'Invalid server response',
      };
    }
  }

  private decodeResponse(statusCode: number, body: string): unknown {
    const payload = this.parsePayload(body);
    if (statusCode >= 200 && statusCode < 300 && payload.success === true) {
      return payload.data;
    }

    const fallback = `Request failed with status ${statusCode}`;
    const message = payload.error === undefined ? fallback : String(payload.error);
    throw new ApiException(message, statusCode);
  }

  async get(route: string, query?: Record<string, string>): Promise<unknown> {
    let response: Response;
    try {
      response = await fetch(this.buildUrl(route, query), {
        method: 'GET',
        headers: await this.headers(),
      });
    } catch (error) {
      throw new ApiException(extractErrorMessage(error));
    }

    const body = await response.text();
    return this.decodeResponse(response.status, body);
  }

  async post(route: string, body?: Record<string, unknown>): Promise<unknown> {
    let response: Response;
    try {
      response = await fetch(this.buildUrl(route), {
        method: 'POST',
        headers: await this.headers(),
        body: JSON.stringify(body ?? {}),
      });
    } catch (error) {
      throw new ApiException(extractErrorMessage(error));
    }

    const payload = await response.text();
    return this.decodeResponse(response.status, payload);
  }

  async postMultipart(
    route: string,
    fields: Record<string, string>,
    fileField: string,
    file: string | MultipartFile,
  ): Promise<unknown> {
    const token = await this.tokenStorage.readToken();
    const headers: Record<string, string> = {};
    if (token !== null && token.length > 0) {
      headers.Authorization = `Bearer ${token}`;
    }

    const fileData = typeof file === 'string' ? { uri: file } : file;
    const hasScheme = /^[a-zA-Z][a-zA-Z0-9+.-]*:\/\//.test(fileData.uri);
    const uri = hasScheme ? fileData.uri : `file://${fileData.uri}`;
    const fileName =
      fileData.name ??
      fileData.uri.split(/[\\/]/).pop()?.split('?')[0] ??
      'upload.bin';
    const fileType =
      fileData.type ??
      (typeof file === 'string'
        ? guessMimeType(file)
        : guessMimeType(fileData.uri));
    const formData = new FormData();
    Object.entries(fields).forEach(([key, value]) => {
      formData.append(key, value);
    });
    formData.append(fileField, {
      uri,
      name: fileName,
      type: fileType,
    } as never);

    let response: Response;
    try {
      response = await fetch(this.buildUrl(route), {
        method: 'POST',
        headers,
        body: formData,
      });
    } catch (error) {
      throw new ApiException(extractErrorMessage(error));
    }

    const payload = await response.text();
    return this.decodeResponse(response.status, payload);
  }
}
